cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/NotificationSubscriberService.java

```diff
@@ -0,0 +1,359 @@
+      }
+    }
+  }
+
+  @Override
+  protected void triggerShutdown() {
+    LOG.info("Stopping NotificationSubscriberService.");
+    for (NotificationSubscriberThread thread : subscriberThreads) {
+      thread.interrupt();
+    }
+
+    LOG.info("NotificationSubscriberService stopped.");
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    if (taskExecutorService != null) {
+      taskExecutorService.shutdownNow();
+    }
+  }
+
+  private abstract class NotificationSubscriberThread extends Thread {
+    private final String topic;
+    private final RetryStrategy scheduleStrategy;
+    protected final Deque<Job> readyJobs;
+    private int emptyFetchCount;
+    private int failureCount;
+    private String messageId;
+
+
+    NotificationSubscriberThread(String topic, @Nullable String messageId) {
+      super(String.format("NotificationSubscriberThread-%s", topic));
+      this.topic = topic;
+      this.messageId = messageId;
+      // TODO: [CDAP-11370] Need to be configured in cdap-default.xml. Retry with delay ranging from 0.1s to 30s
+      scheduleStrategy =
+        co.cask.cdap.common.service.RetryStrategies.exponentialDelay(100, 30000, TimeUnit.MILLISECONDS);
+      this.readyJobs = new ArrayDeque<>();
+    }
+
+    @Override
+    public void run() {
+      while (isRunning()) {
+        try {
+          long sleepTime = processNotifications();
+          // Don't sleep if sleepTime returned is 0
+          if (sleepTime > 0) {
+            TimeUnit.MILLISECONDS.sleep(sleepTime);
+          }
+        } catch (InterruptedException e) {
+          // sleep is interrupted, just exit without doing anything
+        }
+      }
+    }
+
+    /**
+     * Fetch new notifications and update job queue
+     *
+     * @return sleep time in milliseconds before next fetch
+     */
+    private long processNotifications() {
+      try {
+        final MessageFetcher fetcher = messagingContext.getMessageFetcher();
+        emptyFetchCount++;
+        transactional.execute(new TxRunnable() {
+          @Override
+          public void run(DatasetContext context) throws Exception {
+            fetchNotifications(context, fetcher);
+          }
+        });
+      } catch (Exception e) {
+        LOG.warn("Failed to get notification. Will retry in next run", e);
+        failureCount++;
+        // Exponential strategy doesn't use the time component, so doesn't matter what we passed in as startTime
+        return scheduleStrategy.nextRetry(failureCount, 0);
+      }
+      failureCount = 0;
+
+        try {
+          runReadyJobs();
+        } catch (Exception e) {
+          LOG.error("Failed to run scheduled programs", e);
+        }
+      // Back-off if it was empty fetch.
+      if (emptyFetchCount > 0) {
+        return 2000L;
+      }
+      return 0L; // No sleep if the fetch is non-empty
+    }
+
+    private void fetchNotifications(DatasetContext context, MessageFetcher fetcher) throws Exception {
```

Comment:
This is actually more than fetching, right? Better call it `fetchAndProcessNotifications`.